File: server-go/indexer/parser.go
package indexer

import (
	"os"
	"regexp"
	"strings"
)

var (
	// frontmatter 正则
	frontmatterRe = regexp.MustCompile(`(?s)^---\n(.+?)\n---`)
	
	// 文件内定义：【词条】：定义内容
	inlineDefRe = regexp.MustCompile(`【([^】]+)】[：:]\s*(.+?)(?:\n|$)`)
	
	// 策划公式：%% 表达式 %%
	formulaRe = regexp.MustCompile(`%%\s*(.+?)\s*%%`)
	
	// 计算值：[xxx]
	calcValueRe = regexp.MustCompile(`\[([^\]]+)\]`)
	
	// 设计值：<xxx>
	designValueRe = regexp.MustCompile(`<([^>]+)>`)
)

// ParseFile 解析 Markdown 文件
func ParseFile(fullPath, relPath string) ([]*WikiTerm, []*WikiFormula, string) {
	content, err := os.ReadFile(fullPath)
	if err != nil {
		return nil, nil, ""
	}

	text := string(content)
	var terms []*WikiTerm
	var formulas []*WikiFormula
	var scope string

	// 解析 frontmatter
	fm := parseFrontmatter(text)
	if fm != nil {
		scope = fm["scope"]
		
		// 从 frontmatter 创建词条
		term := strings.TrimSuffix(relPath, ".md")
		if idx := strings.LastIndex(term, "/"); idx != -1 {
			term = term[idx+1:]
		}
		
		aliases := []string{term}
		if aliasStr, ok := fm["alias"]; ok {
			for _, a := range strings.Split(aliasStr, ",") {
				a = strings.TrimSpace(a)
				if a != "" && a != term {
					aliases = append(aliases, a)
				}
			}
		}

		// 提取定义（支持 <!-- more --> 截断）
		definition, hasMore := extractDefinition(text)

		if definition != "" {
			terms = append(terms, &WikiTerm{
				Term:           term,
				Aliases:        aliases,
				Definition:     definition,
				Scope:          scope,
				FilePath:       relPath,
				DefinitionType: "file",
				HasMore:        hasMore,
			})
		}
	}

	// 解析文件内定义
	matches := inlineDefRe.FindAllStringSubmatch(text, -1)
	for _, m := range matches {
		termName := m[1]
		def := strings.TrimSpace(m[2])
		terms = append(terms, &WikiTerm{
			Term:           termName,
			Aliases:        []string{termName},
			Definition:     def,
			Scope:          scope,
			FilePath:       relPath,
			DefinitionType: "inline",
		})
	}

	// 解析策划公式
	formulaMatches := formulaRe.FindAllStringSubmatch(text, -1)
	for _, m := range formulaMatches {
		expr := m[1]
		
		var calcValues []string
		for _, cv := range calcValueRe.FindAllStringSubmatch(expr, -1) {
			calcValues = append(calcValues, cv[1])
		}
		
		var designValues []string
		for _, dv := range designValueRe.FindAllStringSubmatch(expr, -1) {
			designValues = append(designValues, dv[1])
		}

		if len(calcValues) > 0 || len(designValues) > 0 {
			formulas = append(formulas, &WikiFormula{
				Expression:       expr,
				CalculatedValues: calcValues,
				DesignValues:     designValues,
				Scope:            scope,
				FilePath:         relPath,
			})
		}
	}

	return terms, formulas, scope
}

func parseFrontmatter(text string) map[string]string {
	match := frontmatterRe.FindStringSubmatch(text)
	if match == nil {
		return nil
	}

	fm := make(map[string]string)
	lines := strings.Split(match[1], "\n")
	for _, line := range lines {
		parts := strings.SplitN(line, ":", 2)
		if len(parts) == 2 {
			key := strings.TrimSpace(parts[0])
			value := strings.TrimSpace(parts[1])
			fm[key] = value
		}
	}
	return fm
}

// moreTagRe 匹配 <!-- more --> 标记
var moreTagRe = regexp.MustCompile(`(?i)^\s*<!--\s*more\s*-->\s*$`)

// extractDefinition 提取定义内容，支持 <!-- more --> 截断
// 返回定义内容和是否有更多内容
func extractDefinition(text string) (string, bool) {
	// 移除 frontmatter
	text = frontmatterRe.ReplaceAllString(text, "")
	
	lines := strings.Split(text, "\n")
	var contentLines []string
	hasMore := false
	
	for _, line := range lines {
		// 检查 <!-- more --> 标记
		if moreTagRe.MatchString(line) {
			hasMore = true
			break
		}
		
		trimmed := strings.TrimSpace(line)
		// 跳过标题、分隔线、空行和内联定义
		if trimmed == "" || strings.HasPrefix(line, "#") || strings.HasPrefix(trimmed, "---") {
			continue
		}
		if inlineDefRe.MatchString(line) {
			continue
		}
		
		contentLines = append(contentLines, line)
	}
	
	return strings.TrimSpace(strings.Join(contentLines, "\n")), hasMore
}

File: server/indexer/parser.ts
/**
 * .md 文件解析器
 * 解析 frontmatter（alias, scope）和正文（文件内定义、策划公式）
 */
import matter from 'gray-matter'
import type { WikiTerm, WikiFormula } from '../../shared/types'

/** 单个文件的解析结果 */
export interface ParseResult {
  /** 文件级词条（来自 frontmatter alias） */
  terms: WikiTerm[]
  /** 文件内定义（【词条】：定义） */
  inlineTerms: WikiTerm[]
  /** 策划公式 */
  formulas: WikiFormula[]
  /** 作用域 */
  scope: string
}

/**
 * 解析单个 .md 文件
 * @param content 文件原始内容
 * @param filePath 相对于 wiki-docs/ 的路径
 */
export function parseFile(content: string, filePath: string): ParseResult {
  const { data: frontmatter, content: markdownContent } = matter(content)

  const scope: string = (frontmatter as Record<string, unknown>).scope as string || ''
  const aliases: string[] = (frontmatter as Record<string, unknown>).alias as string[] || []

  // 计算 frontmatter 占用的行数（用于将正文行号映射回原始文件行号）
  const frontmatterLineCount = getFrontmatterLineCount(content)

  const terms = parseAliasTerms(markdownContent, filePath, scope, aliases, frontmatterLineCount)
  const inlineTerms = parseInlineTerms(markdownContent, filePath, scope, frontmatterLineCount)
  const formulas = parseFormulas(markdownContent, filePath, scope, frontmatterLineCount)

  return { terms, inlineTerms, formulas, scope }
}

/**
 * 计算 frontmatter 占用的行数（包含两行 ---）
 * 如果没有 frontmatter 则返回 0
 */
function getFrontmatterLineCount(content: string): number {
  const trimmed = content.trimStart()
  if (!trimmed.startsWith('---')) return 0
  const endIndex = trimmed.indexOf('---', 3)
  if (endIndex === -1) return 0
  const frontmatterBlock = trimmed.slice(0, endIndex + 3)
  return frontmatterBlock.split('\n').length
}

/**
 * 从 frontmatter alias 解析文件级词条
 * 使用全文（去除标题/空行/内联定义行）作为定义内容
 * 支持 <!-- more --> 标记截断摘要
 */
function parseAliasTerms(
  content: string,
  filePath: string,
  scope: string,
  aliases: string[],
  frontmatterLineCount: number,
): WikiTerm[] {
  if (aliases.length === 0) return []

  const lines = content.split('\n')
  // 找到第一行有效内容的行号（用于定位）
  let firstContentLine = -1
  const contentLines: string[] = []
  let hasMore = false
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    
    // 检查 <!-- more --> 标记（支持多种格式）
    if (line.trim().match(/^<!--\s*more\s*-->$/i)) {
      hasMore = true
      break
    }
    
    if (
      !line.startsWith('#') &&
      !line.startsWith('---') &&
      line.trim().length > 0 &&
      !line.match(/【[^】]+】：/)
    ) {
      contentLines.push(line)
      if (firstContentLine === -1) {
        firstContentLine = i
      }
    }
  }
  const definition = contentLines.join('\n').trim()
  if (!definition) return []

  const term: WikiTerm = {
    term: aliases[0],
    aliases,
    definition,
    scope,
    filePath,
    definitionType: 'file',
    line: firstContentLine >= 0 ? frontmatterLineCount + firstContentLine + 1 : undefined,
    hasMore,
  }

  return [term]
}

/**
 * 解析文件内定义：【词条】：定义内容
 * 跳过代码块内的内容
 */
function parseInlineTerms(
  content: string,
  filePath: string,
  scope: string,
  frontmatterLineCount: number,
): WikiTerm[] {
  const terms: WikiTerm[] = []
  const lines = content.split('\n')
  let inFence = false
  let fenceMarker = ''

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    // 处理围栏代码块
    if (toggleFence(line.trim(), { inFence, fenceMarker })) {
      inFence = !inFence
      if (inFence) {
        const match = line.trim().match(/^([`~]{3,})/)
        fenceMarker = match ? match[1][0] : ''
      } else {
        fenceMarker = ''
      }
      continue
    }
    if (inFence) continue

    // 移除行内代码后匹配
    const cleaned = line.replace(/`[^`]*`/g, '')
    const match = cleaned.match(/【([^】]+)】：(.+)/)
    if (match) {
      terms.push({
        term: match[1],
        aliases: [match[1]],
        definition: match[2].trim(),
        scope,
        filePath,
        definitionType: 'inline',
        line: frontmatterLineCount + i + 1,
      })
    }
  }

  return terms
}

/**
 * 解析策划公式：%% 表达式 %%
 * 提取计算值 [xxx] 和设计值 <xxx>
 */
function parseFormulas(
  content: string,
  filePath: string,
  scope: string,
  frontmatterLineCount: number,
): WikiFormula[] {
  const formulas: WikiFormula[] = []
  const lines = content.split('\n')
  let inFence = false
  let fenceMarker = ''

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    if (toggleFence(line.trim(), { inFence, fenceMarker })) {
      inFence = !inFence
      if (inFence) {
        const match = line.trim().match(/^([`~]{3,})/)
        fenceMarker = match ? match[1][0] : ''
      } else {
        fenceMarker = ''
      }
      continue
    }
    if (inFence) continue

    const cleaned = line.replace(/`[^`]*`/g, '')
    const formulaRegex = /%%\s*([^%]+)\s*%%/g
    let match
    while ((match = formulaRegex.exec(cleaned)) !== null) {
      const expression = match[1].trim()
      formulas.push({
        expression,
        calculatedValues: extractCalculatedValues(expression),
        designValues: extractDesignValues(expression),
        scope,
        filePath,
        line: frontmatterLineCount + i + 1,
      })
    }
  }

  return formulas
}

/** 提取计算值 [xxx] */
export function extractCalculatedValues(formula: string): string[] {
  const matches: string[] = []
  const regex = /\[([^\]]+)\]/g
  let match
  while ((match = regex.exec(formula)) !== null) {
    matches.push(match[1])
  }
  return matches
}

/** 提取设计值 <xxx> */
export function extractDesignValues(formula: string): string[] {
  const matches: string[] = []
  const regex = /<([^>]+)>/g
  let match
  while ((match = regex.exec(formula)) !== null) {
    matches.push(match[1])
  }
  return matches
}

/**
 * 检测围栏代码块的开始/结束
 * 返回 true 表示当前行是围栏标记行
 */
function toggleFence(
  line: string,
  state: { inFence: boolean; fenceMarker: string },
): boolean {
  const fenceMatch = line.match(/^([`~]{3,})/)
  if (!fenceMatch) return false

  const marker = fenceMatch[1][0]
  if (!state.inFence) {
    return true // 进入围栏
  } else if (marker === state.fenceMarker) {
    return true // 离开围栏
  }
  return false
}
